web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java

```diff
@@ -83,22 +96,34 @@
  */
 public class TokenBasedRememberMeServices extends AbstractRememberMeServices {
 
+	private RememberMeHashingAlgorithm hashingAlgorithm;
+
 	public TokenBasedRememberMeServices(String key, UserDetailsService userDetailsService) {
+		this(key, userDetailsService, RememberMeHashingAlgorithm.UNSET);
+	}
+
+	/**
+	 * @since 5.5
+	 */
+	public TokenBasedRememberMeServices(String key, UserDetailsService userDetailsService,
+			RememberMeHashingAlgorithm hashingAlgorithm) {
 		super(key, userDetailsService);
+		this.hashingAlgorithm = hashingAlgorithm;
 	}
 
 	@Override
 	protected UserDetails processAutoLoginCookie(String[] cookieTokens, HttpServletRequest request,
 			HttpServletResponse response) {
-		if (cookieTokens.length != 3) {
+		if (!(cookieTokens.length == 3 || cookieTokens.length == 4)) {
```

Comment:
I think we can simplify the checks here and create an util class:

```java
private static class CookieTokens {

		private final String username;

		private final long expirationTime;

		private final RememberMeHashingAlgorithm algorithm;

		private final String signature;

		private CookieTokens(String[] cookieTokens) {
			if (!(cookieTokens.length == 3 || cookieTokens.length == 4)) {
				throw new InvalidCookieException(
						"Cookie token did not contain 3 or 4 tokens, but contained '" + Arrays.asList(cookieTokens) + "'");
			}
			this.username = getUsername(cookieTokens);
			this.expirationTime = getTokenExpiryTime(cookieTokens);
			this.algorithm = getTokenHashingAlgorithm(cookieTokens);
			this.signature = getSignature(cookieTokens);
		}

		private String getUsername(String[] cookieTokens) {
			return cookieTokens[0];
		}

		private long getTokenExpiryTime(String[] cookieTokens) {
			try {
				return Long.parseLong(cookieTokens[1]);
			}
			catch (NumberFormatException nfe) {
				throw new InvalidCookieException(
						"Cookie token[1] did not contain a valid number (contained '" + cookieTokens[1] + "')");
			}
		}

		private RememberMeHashingAlgorithm getTokenHashingAlgorithm(String[] cookieTokens) {
			if (cookieTokens.length == 3) {
				return null;
			}
			return RememberMeHashingAlgorithm.findByIdentifier(cookieTokens[2]);
		}

		private String getSignature(String[] cookieTokens) {
			return (cookieTokens.length == 4) ? cookieTokens[3] : cookieTokens[2];
		}

	}
```

And this method should use the properties from this class:

```java
CookieTokens properties = new CookieTokens(cookieTokens);
if (isTokenExpired(properties.expirationTime)) {
	throw new InvalidCookieException("Cookie token[1] has expired (expired on '" + new Date(properties.expirationTime)
			+ "'; current time is '" + new Date() + "')");
}
...
```